### PR TITLE
zh: add Advanced StatefulSet

### LINKED ISSUE #9: add docs about Advanced Statefulset feature
this is a track issue for docs about Advanced StatefulSet feature

### PR DESCRIPTION
**Pull request by cofyc**
### What is changed, added or deleted? (Required)

fixes #9 

### Which TiDB Operator version(s) do your changes apply to? (Required)

- [x] master (the latest development version, including v1.1 changes for now)
- [ ] v1.0 (TiDB Operator 1.0 versions)

### What is the related PR or file link(s)?

- This PR is translated from: 
- Other reference link(s):

### CODE REVIEW

File: zh/advanced-statefulset.md [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    config: {}
+  tidb:
+    baseImage: pingcap/tidb
+    replicas: 2
+    service:
+      type: ClusterIP
+    config: {}
+```
+
+会部署 4 个 tikv 实例，分别为 basic-tikv-0, basic-tikv-1, ..., basic-tikv-3。若想缩容掉 basic-tikv-1 需要在修改 `spec.tikv.replicas` 为 3 同时配置以下 annotations:
```
Comment (DanielZhangQD) [inline zh/advanced-statefulset.md]: Suggest adding scaling out cases:

- scaling out one pod that has been deleted before
- scaling out one pod that has not been deleted before

Comment (DanielZhangQD) [reply zh/advanced-statefulset.md]: May also include PVC operations.

File: zh/advanced-statefulset.md [hunk 2]
```diff
@@ -0,0 +1,156 @@
+---
+title: 使用增强型 StatefulSet 控制器
+summary: 介绍如何开启、使用增强型 StatefulSet 控制器
+category: alpha
+---
+
+# 使用增强型 StatefulSet 控制器
+
+Kubernetes 内置 [StatefulSet](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/) 为 Pods 分配连续的序号。比如 3 个副本时，Pods 分别为 pod-0, pod-1, pod-2 。扩缩容时，必须在尾部增加或删除 Pods 。比如扩容到 4 个副本时，会新增 pod-3 。缩容到 2 副本时，会删除 pod-2 。
```
Comment (ran-huang) [inline zh/advanced-statefulset.md]: No spaces between English words and punctuation. Please also update the rest of this document.
```suggestion
Kubernetes 内置 [StatefulSet](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/) 为 Pods 分配连续的序号。比如 3 个副本时，Pods 分别为 pod-0, pod-1, pod-2。扩缩容时，必须在尾部增加或删除 Pods。比如扩容到 4 个副本时，会新增 pod-3。缩容到 2 副本时，会删除 pod-2。 
```

File: zh/advanced-statefulset.md [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+1. 载入 Advanced StatefulSet 的 CRD 文件：
+
+    Kubernetes 1.16 之前版本：
+
+    {{< copyable "shell-regular" >}}
+
+    ```shell
+    kubectl apply -f https://raw.githubusercontent.com/pingcap/tidb-operator/master/manifests/advanced-statefulset-crd.v1beta1.yaml
+    ```
```
Comment (ran-huang) [inline zh/advanced-statefulset.md]:
```suggestion
    * Kubernetes 1.16 之前版本：

        {{< copyable "shell-regular" >}}

        ```shell
        kubectl apply -f https://raw.githubusercontent.com/pingcap/tidb-operator/master/manifests/advanced-statefulset-crd.v1beta1.yaml
        ```
```

File: zh/advanced-statefulset.md [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+    Kubernetes 1.16 之后:
+
+    {{< copyable "shell-regular" >}}
+
+    ```
+    kubectl apply -f https://raw.githubusercontent.com/pingcap/tidb-operator/master/manifests/advanced-statefulset-crd.v1.yaml
+    ```
```
Comment (ran-huang) [inline zh/advanced-statefulset.md]: ditto
```suggestion
    * Kubernetes 1.16 之后:

        {{< copyable "shell-regular" >}}

        ```
        kubectl apply -f https://raw.githubusercontent.com/pingcap/tidb-operator/master/manifests/advanced-statefulset-crd.v1.yaml
        ```
```

File: zh/advanced-statefulset.md [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    {{< copyable "shell-regular" >}}
+
+    ```yaml
+    features:
+    - AdvancedStatefulSet=true
+    advancedStatefulset:
+      create: true
+    ```
+
+    然后升级 TiDB Operator ，具体可参考[升级 TiDB Operator 文档]((upgrade-TiDB Operator.md))。
```
Comment (ran-huang) [inline zh/advanced-statefulset.md]:
```suggestion
    然后升级 TiDB Operator，具体可参考[升级 TiDB Operator 文档](upgrade-TiDB Operator.md)。
```

File: zh/advanced-statefulset.md [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+> **注意：**
+>
+> TiDB Operator 通过开启 `AdvancedStatefulSet` 特性，会将当前 `StatefulSet`
+> 对象转换成 `AdvancedStatefulSet 对象。但并不支持关闭 `AdvancedStatefulSet`
+> 特性后，自动从 `AdvancedStatefulSet` 转换为 Kubernetes 内置的 `StatefulSet`
+> 对象。
```
Comment (ran-huang) [inline zh/advanced-statefulset.md]:
```suggestion
> TiDB Operator 通过开启 `AdvancedStatefulSet` 特性，会将当前 `StatefulSet` 对象转换成 `AdvancedStatefulSet 对象。但是，TiDB Operator 不支持在关闭 `AdvancedStatefulSet` 特性后，自动从 `AdvancedStatefulSet` 转换为 Kubernetes 内置的 `StatefulSet` 对象。
```

File: zh/advanced-statefulset.md [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+{{< copyable "shell-regular" >}}
+
+```shell
+kubectl get -n <namespace> asts
+```
+
+### 操作 TidbCluster 对象指定 pod 进行缩容
+
+使用增强型 StatefulSet 时，在对 TidbCluster 进行缩容时，除了减少副本数，可同时通过配置 annotations 指定对 pd, tidb 或 tikv 组件下任意一个 pod 进行缩容。
```
Comment (ran-huang) [inline zh/advanced-statefulset.md]:
```suggestion
使用增强型 StatefulSet 时，在对 TidbCluster 进行缩容时，除了减少副本数，可同时通过配置 annotations 指定对 PD，TiDB 或 TiKV 组件下任意一个 Pod 进行缩容。
```

File: zh/advanced-statefulset.md (hunk 1 above)
Comment (ran-huang) [inline zh/advanced-statefulset.md]:
```suggestion
上述配置会部署 4 个 tikv 实例，分别为 basic-tikv-0，basic-tikv-1，...，basic-tikv-3。若想缩容掉 basic-tikv-1 需要修改 `spec.tikv.replicas` 为 3，同时配置以下 annotations:
```

File: zh/advanced-statefulset.md [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+    service:
+      type: ClusterIP
+    config: {}
+```
+
+支持的 annotations 为：
+
+- `pd.tidb.pingcap.com/delete-slots`: 指定 pd 组件需要删除的 pod 序号。
+- `tidb.tidb.pingcap.com/delete-slots`: 指定 tidb 组件需要删除的 pod 序号
+- `tikv.tidb.pingcap.com/delete-slots`: 指定 tikv 组件需要删除的 pod 序号
```
Comment (ran-huang) [inline zh/advanced-statefulset.md]: Please use Chinese colons when there are both Chinese and English words in the text.
```suggestion
- `pd.tidb.pingcap.com/delete-slots`：指定 PD 组件需要删除的 Pod 序号。
- `tidb.tidb.pingcap.com/delete-slots`：指定 TiDB 组件需要删除的 Pod 序号。
- `tikv.tidb.pingcap.com/delete-slots`：指定 TiKV 组件需要删除的 Pod 序号。
```

**Comment by ran-huang** [comment]
@cofyc Please resolve the conflicts.

File: zh/TOC.md [hunk 9]
```diff
@@ -61,6 +61,7 @@
   - [tkctl](use-tkctl.md)
   - [TiDB Toolkit](tidb-toolkit.md)
 + 组件
-    - [TiDB Scheduler 扩展调度器](tidb-scheduler.md)
```
Comment (cofyc) [inline zh/TOC.md]: fixed indent

File: zh/advanced-statefulset.md [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+
+    * Kubernetes 1.16 之后:
+
+        {{< copyable "shell-regular" >}}
+
+        ```
+        kubectl apply -f https://raw.githubusercontent.com/pingcap/tidb-operator/master/manifests/advanced-statefulset-crd.v1.yaml
+        ```
+
+2. 在 TiDB Operator chart 的 values 中启用 `AdvancedStatefulSet` 特性：
```
Comment (ran-huang) [inline zh/advanced-statefulset.md]: Is this what you mean?
```suggestion
2. 在 TiDB Operator chart 的 `values.yaml` 中启用 `AdvancedStatefulSet` 特性：
```

File: zh/advanced-statefulset.md [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+> **注意：**
+>
+> TiDB Operator 通过开启 `AdvancedStatefulSet` 特性，会将当前 `StatefulSet` 对象转换成 `AdvancedStatefulSet 对象。但是，TiDB Operator 不支持在关闭 `AdvancedStatefulSet` 特性后，自动从 `AdvancedStatefulSet` 转换为 Kubernetes 内置的 `StatefulSet` 对象。
+
+## 使用
+
+### 通过 kubectl 查看 AdvancedStatefulSet 对象
+
+`AdvancedStatefulSet` 数据格式与 `StatefulSet` 完全一致，但以 CRD
+方式实现，别名为 `asts` ，可通过一下方法查看命令空间下的对象。
```
Comment (ran-huang) [inline zh/advanced-statefulset.md]:
```suggestion
`AdvancedStatefulSet` 数据格式与 `StatefulSet` 完全一致，但以 CRD 方式实现，别名为 `asts` ，可通过以下方法查看命令空间下的对象。
```

File: zh/advanced-statefulset.md [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+    config: {}
+  tidb:
+    baseImage: pingcap/tidb
+    replicas: 2
+    service:
+      type: ClusterIP
+    config: {}
+```
+
+上述配置会部署 4 个 tikv 实例，分别为 basic-tikv-0，basic-tikv-1，...，basic-tikv-3。若想缩容掉 basic-tikv-1 需要修改 `spec.tikv.replicas` 为 3，同时配置以下 annotations:
```
Comment (ran-huang) [inline zh/advanced-statefulset.md]:
```suggestion
上述配置会部署 4 个 TiKV 实例，分别为 basic-tikv-0，basic-tikv-1，...，basic-tikv-3。若想缩容掉 basic-tikv-1 需要修改 `spec.tikv.replicas` 为 3，同时配置以下 annotations:
```

File: zh/advanced-statefulset.md [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+
+{{< copyable "" >}}
+
+```yaml
+metadata:
+  annotations:
+    tikv.tidb.pingcap.com/delete-slots: '[1]'
+```
+
+> **注意**：对 replicas 和 delete slot annotation 的修改需在同一个操作中完成，不然控制器会根据修改一般的期望进行操作。
```
Comment (ran-huang) [inline zh/advanced-statefulset.md]:
```suggestion
> **注意：**
>
> 对 `replicas` 和 `delete slot annotation` 的修改需在同一个操作中完成，不然控制器会根据修改一般的期望进行操作。
```

File: zh/advanced-statefulset.md [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+- `tidb.tidb.pingcap.com/delete-slots`：指定 TiDB 组件需要删除的 Pod 序号。
+- `tikv.tidb.pingcap.com/delete-slots`：指定 TiKV 组件需要删除的 Pod 序号。
+
+其中 Annotation 值为 JSON 的整数数组，比如 `[0]`, `[0,1]`, `[1,3]` 等。
+
+### 操作 TidbCluster 对象在指定位置进行扩容
+
+对前面缩容进行反向操作，即可恢复 pod-1 。
+
+> **注意**：同常规 StatefulSet 缩容一样，并不会主动删除 Pod 关联的 PVC ，若想避免使用之前数据，在原位置处扩容，需主动删除关联的 PVC 。
```
Comment (ran-huang) [inline zh/advanced-statefulset.md]:
```suggestion
> **注意：**
>
> 同常规 StatefulSet 缩容一样，并不会主动删除 Pod 关联的 PVC，若想避免使用之前数据，在原位置处扩容，需主动删除关联的 PVC。
```

File: zh/advanced-statefulset.md [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+    config: {}
+  tidb:
+    baseImage: pingcap/tidb
+    replicas: 2
+    service:
+      type: ClusterIP
+    config: {}
+```
+
+其中 delete slots annotations 可留空或完全删除，均可。
```
Comment (ran-huang) [inline zh/advanced-statefulset.md]:
```suggestion
其中 delete slots annotations 可留空，也可完全删除。
```

File: zh/advanced-statefulset.md (hunk 10 above)
Comment (cofyc) [reply zh/advanced-statefulset.md]: yes, thanks!

File: zh/advanced-statefulset.md (hunk 1 above)
Comment (cofyc) [reply zh/advanced-statefulset.md]: done

File: zh/advanced-statefulset.md [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+
+```yaml
+metadata:
+  annotations:
+    tikv.tidb.pingcap.com/delete-slots: '[1]'
+```
+
+> **注意：**
+>
+> 对 `replicas` 和 `delete slot annotation` 的修改需在同一个操作中完成，不然控制器会根据修改一般的期望进行操作。
```
Comment (DanielZhangQD) [inline zh/advanced-statefulset.md]:
```suggestion
> 对 `replicas` 和 `delete-slots annotation` 的修改需在同一个操作中完成，不然控制器会根据修改一般的期望进行操作。
```

File: zh/advanced-statefulset.md [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+
+- `pd.tidb.pingcap.com/delete-slots`：指定 PD 组件需要删除的 Pod 序号。
+- `tidb.tidb.pingcap.com/delete-slots`：指定 TiDB 组件需要删除的 Pod 序号。
+- `tikv.tidb.pingcap.com/delete-slots`：指定 TiKV 组件需要删除的 Pod 序号。
+
+其中 Annotation 值为 JSON 的整数数组，比如 `[0]`, `[0,1]`, `[1,3]` 等。
+
+### 操作 TidbCluster 对象在指定位置进行扩容
+
+对前面缩容进行反向操作，即可恢复 pod-1 。
```
Comment (DanielZhangQD) [inline zh/advanced-statefulset.md]:
```suggestion
对前面缩容进行反向操作，即可恢复 basic-tikv-1。
```

File: zh/advanced-statefulset.md [hunk 18]
```diff
@@ -0,0 +1,203 @@
...
+
+> **注意：**
+>
+> 同常规 StatefulSet 缩容一样，并不会主动删除 Pod 关联的 PVC，若想避免使用之前数据，在原位置处扩容，需主动删除关联的 PVC。
```
Comment (DanielZhangQD) [inline zh/advanced-statefulset.md]:
```suggestion
> 同常规 StatefulSet 缩容一样，并不会主动删除 Pod 关联的 PVC，若想避免使用之前数据，在原位置处扩容之前，需主动删除之前关联的 PVC。
```

File: zh/advanced-statefulset.md [hunk 19]
```diff
@@ -0,0 +1,203 @@
...
+    config: {}
+  tidb:
+    baseImage: pingcap/tidb
+    replicas: 2
+    service:
+      type: ClusterIP
+    config: {}
+```
+
+其中 delete slots annotations 可留空，也可完全删除。
```
Comment (DanielZhangQD) [inline zh/advanced-statefulset.md]:
```suggestion
其中 delete-slots annotations 可留空，也可完全删除。
```

File: zh/advanced-statefulset.md [hunk 20]
```diff
@@ -0,0 +1,203 @@
...
+
+支持的 annotations 为：
+
+- `pd.tidb.pingcap.com/delete-slots`：指定 PD 组件需要删除的 Pod 序号。
+- `tidb.tidb.pingcap.com/delete-slots`：指定 TiDB 组件需要删除的 Pod 序号。
+- `tikv.tidb.pingcap.com/delete-slots`：指定 TiKV 组件需要删除的 Pod 序号。
+
+其中 Annotation 值为 JSON 的整数数组，比如 `[0]`, `[0,1]`, `[1,3]` 等。
+
+### 操作 TidbCluster 对象在指定位置进行扩容
```
Comment (DanielZhangQD) [inline zh/advanced-statefulset.md]: Should we add a section for the normal scaling out, which is the same as the k8s sts controller?

Comment (cofyc) [reply zh/advanced-statefulset.md]: I don't think it's necessary because there is no difference.

File: zh/advanced-statefulset.md [hunk 21]
```diff
@@ -0,0 +1,203 @@
...
+
+在使用本地存储时，Pods 与 Nodes 存储资源绑定，无法自由调度。若希望删除掉中间某个 Pod ，以便维护其所在的 Node 但并没有其他 Node 可以迁移时，或者某个 Pod 故障想直接删除，另起一个序号不一样的 Pod 时，无法通过内置 StatefulSet 实现。
+
+[增强型 StatefulSet 控制器](https://github.com/pingcap/advanced-statefulset) 基于内置 StatefulSet 实现，新增了自由控制 Pods 序号的功能。本文介绍如何在 TiDB Operator 中使用。
+
+## 开启
+
+1. 载入 Advanced StatefulSet 的 CRD 文件：
+
+    * Kubernetes 1.16 之前版本：
```
Comment (ran-huang) [inline zh/advanced-statefulset.md]: @cofyc K8s version <= 1.16 or < 1.16?

File: zh/advanced-statefulset.md [hunk 22]
```diff
@@ -0,0 +1,203 @@
...
+    {{< copyable "shell-regular" >}}
+
+    ```yaml
+    features:
+    - AdvancedStatefulSet=true
+    advancedStatefulset:
+      create: true
+    ```
+
+    然后升级 TiDB Operator，具体可参考[升级 TiDB Operator 文档](upgrade-TiDB Operator.md)。
```
Comment (ran-huang) [inline zh/advanced-statefulset.md]: deadlink

File: zh/advanced-statefulset.md [hunk 23]
```diff
@@ -0,0 +1,203 @@
...
+
+> **注意：**
+>
+> TiDB Operator 通过开启 `AdvancedStatefulSet` 特性，会将当前 `StatefulSet` 对象转换成 `AdvancedStatefulSet 对象。但是，TiDB Operator 不支持在关闭 `AdvancedStatefulSet` 特性后，自动从 `AdvancedStatefulSet` 转换为 Kubernetes 内置的 `StatefulSet` 对象。
```
Comment (ran-huang) [inline zh/advanced-statefulset.md]: missing a ``` ` ```.

File: zh/advanced-statefulset.md [hunk 24]
```diff
@@ -0,0 +1,203 @@
...
+
+## 使用
+
+### 通过 kubectl 查看 AdvancedStatefulSet 对象
+
+`AdvancedStatefulSet` 数据格式与 `StatefulSet` 完全一致，但以 CRD 方式实现，别名为 `asts` ，可通过以下方法查看命令空间下的对象。
```
Comment (ran-huang) [inline zh/advanced-statefulset.md]: 命名空间

File: zh/advanced-statefulset.md [hunk 25]
```diff
@@ -0,0 +1,203 @@
...
+
+    * Kubernetes 1.16 之前版本：
+
+        {{< copyable "shell-regular" >}}
+
+        ```shell
+        kubectl apply -f https://raw.githubusercontent.com/pingcap/tidb-operator/master/manifests/advanced-statefulset-crd.v1beta1.yaml
+        ```
+
+    * Kubernetes 1.16 之后:
```
Comment (ran-huang) [inline zh/advanced-statefulset.md]: 1.16 及之后版本

File: zh/advanced-statefulset.md (hunk 24 above)
Comment (ran-huang) [reply zh/advanced-statefulset.md]: fixed in https://github.com/pingcap/docs-tidb-operator/pull/110

File: zh/advanced-statefulset.md (hunk 23 above)
Comment (ran-huang) [reply zh/advanced-statefulset.md]: fixed in https://github.com/pingcap/docs-tidb-operator/pull/110

File: zh/advanced-statefulset.md (hunk 22 above)
Comment (ran-huang) [reply zh/advanced-statefulset.md]: fixed in https://github.com/pingcap/docs-tidb-operator/pull/110

File: zh/advanced-statefulset.md (hunk 25 above)
Comment (ran-huang) [reply zh/advanced-statefulset.md]: fixed in https://github.com/pingcap/docs-tidb-operator/pull/110

**Pull request merged**